### ISSUE TITLE
Technical type should be checked to make mistakes easier to diagnose.

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#1199]
# What happened
Han was using RAP on his local machine, trying to compile "[Arbeidsduur.adl](https://github.com/AmpersandTarski/Ampersand/blob/2aea634743a21f798d5cda8f10a068374eb85f83/testing/Travis/testcases/Misc/Arbeidsduur.adl)".
After compiling Arbeidsduur for over 30 seconds, he got unexpected errors:
![image](https://user-images.githubusercontent.com/275768/128880379-5a0dec60-ee49-4755-926a-d0f381ee4d68.png)
I can reproduce these errors, which is how I got this screenshot.

# What I expected
Arbeidsduur is a correct script. It should work.

# Version of Ampersand that was used
I used the Ampersand container `ampersandtarski/ampersand:development`, which runs commit 6744ff50054fdfcfbb6cd64f71baf06acb28ac25.

# Steps to reproduce

1. git clone https://github.com/AmpersandTarski/RAPinstall
2. cd RAPinstall
3. docker network create proxy
4. docker compose up -d

# Screenshot / Video
Here is [Han's video](https://ordinaweb.sharepoint.com/:v:/r/sites/Ampersand521/Gedeelde%20documenten/General/Recordings/Vergadering%20in%20General-20210808_134516-Opname%20van%20vergadering.mp4?csf=1&web=1&e=XsB7xl) of this event. (15 minutes)

**Comment by stefjoosten** [comment]
# Analysis
Here is an excerpt from the log file. It shows the moment that the database produces the first error:
```log
rap4  | [2021-08-10 13:30:38] CORE.INFO: Atom added to concept: isDirecteVoorloperVa...zoek] is irreflexief[Text] [] {"request_id":"56e85ffa3d","ip":"192.168.32.2","method":"PATCH","url":"/api/v1/resource/Script/Script_1001417f-c9d9-499f-833c-cd40653a72c5/Nieuw_32_script"}
rap4  | [2021-08-10 13:30:38] CORE.INFO: Atom added to concept: De nieuwe arbeidsduu... aanpassing blijken.[Text] [] {"request_id":"56e85ffa3d","ip":"192.168.32.2","method":"PATCH","url":"/api/v1/resource/Script/Script_1001417f-c9d9-499f-833c-cd40653a72c5/Nieuw_32_script"}
rap4  | [2021-08-10 13:30:38] DATABASE.ERROR: Data too long for column 'Text' at row 1 [] {"request_id":"56e85ffa3d","ip":"192.168.32.2","method":"PATCH","url":"/api/v1/resource/Script/Script_1001417f-c9d9-499f-833c-cd40653a72c5/Nieuw_32_script"}
```
So here is my hypothesis. The concept `Text` cannot cope with the data it is fed, which breaks the database.

Further digging showed that the file `Arbeidsduur.adl` contains the statement
```Ampersand
REPRESENT Text TYPE ALPHANUMERIC
```
I have changed this to
```Ampersand
REPRESENT Text TYPE BIGALPHANUMERIC
```
Now the thing works:
![image](https://user-images.githubusercontent.com/275768/128883829-a53dda78-9445-40cf-af9f-e4a284902751.png)

# Diagnosis
Once again this is a mistake of a database field that is of the wrong technical type.
Again, I have spent hours (this time approx 5 to diagnose this particular bug. It is very hard to diagnose database errors, so this takes tons of valuable time. We should do something to make diagnosis simpler.

**Comment by hanjoosten** [comment]
I am surprised that your solution works. I would rather think that this was caused by the database not exepting the contents of the scriptversion. That content is stored in the database as wel, I suppose. I had a brief look, but I couldn't find it within a couple of minutes. But the error in the log file clearly states that the error occurs when storing the script into the database. 

In RAP, there is also a Concept named `Text`. It is defined as ALPHANUMERIC. (line 4 of Analyse.ifc). That smells a dirty workaround for the fact that the markup of a purpose isn't well specified in FormalAmpersand.adl ? I think that @stefjoosten  still has work to do.

**Comment by hanjoosten** [comment]
A note on the part that these errors are hard to diagnose: Sure, you are absolutely right. However, this is not something that ampersand can solve. This is runtime behaviour: At runtime, a population is imported into the database. This population is generated, but at that time I don't think that there is knowledge of the REPRESENTations of the concepts. 
To diagnose this deeper, we need to know exactly what causes this error. I think the lenght of the script is the culprit, but you seem to have shown that such is not the case.

**Comment by Michiel-s** [comment]
> A note on the part that these errors are hard to diagnose: Sure, you are absolutely right. However, this is not something that ampersand can solve. This is runtime behaviour: At runtime, a population is imported into the database. This population is generated, but at that time I don't think that there is knowledge of the REPRESENTations of the concepts. 
> 
> To diagnose this deeper, we need to know exactly what causes this error. I think the lenght of the script is the culprit, but you seem to have shown that such is not the case. 

I do think the length of the script (in combination with the small REPRESENTation). That is why the issue was solved after changing it from ALPHANUMERIC to BIGALPHANUMERIC. The former is max 255 chars. The latter has the size of type Text in the database.

**Comment by stefjoosten** [comment]
> HJ wrote:
> I am surprised that your solution works. I would rather think that this was caused by the database not accepting the contents of the script version. That content is stored in the database as well, I suppose. I had a brief look, but I couldn't find it within a couple of minutes. But the error in the log file clearly states that the error occurs when storing the script into the database.

SJ: In my analysis, I excluded the possibility that the script was too large for the database to contain. I have seen the entire script in the appropriate database field.

> HJ wrote:
> In RAP, there is also a Concept named `Text`. It is defined as ALPHANUMERIC. (line 4 of Analyse.ifc). That smells like a dirty workaround for the fact that the markup of a purpose isn't well specified in FormalAmpersand.adl. I think that @stefjoosten still has work to do.

SJ: I agree. However, that is another issue than this one.

**Comment by stefjoosten** [comment]
> hanjoosten wrote:
> A note on the part that these errors are hard to diagnose: Sure, you are absolutely right. However, this is not something that ampersand can solve. This is runtime behaviour: At runtime, a population is imported into the database. This population is generated, but at that time I don't think that there is knowledge of the REPRESENTations of the concepts.

Of course, Ampersand cannot diagnose. But it can most certainly be more helpful. Maybe the prototype framework
can produce a message before the database runs into an error. A message from the prototype framework can be much more informative. I am not sure that Ampersand can generate a useful message at compile time because the problem occurs at runtime. However, it might communicate a condition that the prototype framework can then check at runtime. So, there should be ample opportunity to prevent this error from hitting the database.

> To diagnose this deeper, we need to know exactly what causes this error. I think the lenght of the script is the culprit, but you seem to have shown that such is not the case.

This particular incident was caused by the fact that an atom at runtime did not satisfy the constraints of its technical type, thus causing the database to produce an error.
